### PR TITLE
Feature: injector flux diagnostics and plots 

### PR DESCRIPTION
**Pull request by KZhu-ME**
Can be initiated using diode setup.

### CODE REVIEW

File: mewarpx/mewarpx/appendablearray.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            if len(data.shape) == len(self._unitshape): lendata = 1
+            else:                                       lendata = data.shape[0]
+        self._extend(lendata)
+        newlen = self._datalen + lendata
+        self._array[self._datalen:newlen,...] = data
+        self._datalen = newlen
+
+    def data(self):
+        """
+    Return the data.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Return the data.
```

File: mewarpx/mewarpx/appendablearray.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        return self._array[:len(self),...]
+
+    def setautobump(self,a):
+        """
+    Set the autobump attribute to the value specified.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Set the autobump attribute to the value specified.
```

File: mewarpx/mewarpx/appendablearray.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        self._autobump = a
+
+    def getautobump(self):
+        """
+    Get the autobump attribute
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Get the autobump attribute
```

File: mewarpx/mewarpx/appendablearray.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        return self._autobump
+
+    def cleardata(self):
+        """
+    Reset the array so it has a length of zero.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Reset the array so it has a length of zero.
```

File: mewarpx/mewarpx/appendablearray.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        self._datalen = 0
+
+    def resetdata(self, data):
+        """
+    Resets the data to be the input values - all of the original data is thrown
+    away. The unit shape of data must be the same.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Resets the data to be the input values - all of the original data is thrown
        away. The unit shape of data must be the same.
```

File: mewarpx/mewarpx/appendablearray.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+            return (1,)
+        else:
+            return self._unitshape
+
+    def reshape(self,newunitshape):
+        """
+    Change the shape of the appendable unit. Can only be used if a unitshape was
+    specified on creation.
+     - newunitshape: must have the same number of dimensions as the original
+                     unitshape
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
        Change the shape of the appendable unit. Can only be used if a unitshape was
        specified on creation.
         - newunitshape: must have the same number of dimensions as the original
                         unitshape
```

File: mewarpx/mewarpx/emission.py [hunk 7]
```diff
@@ -107,10 +109,9 @@ def init_injectedparticles(self, fieldlist):
...
-        raise NotImplementedError
-        # self._injectedparticles_fields = fieldlist
-        # self._injectedparticles_data = warp.AppendableArray(
-        #     typecode='d', unitshape=[len(fieldlist)])
+        self._injectedparticles_fields = fieldlist
+        self._injectedparticles_data = appendablearray.AppendableArray(
+            typecode='d', unitshape=[len(fieldlist)])
 
     def record_injectedparticles(self, species, E_total=None, w=None,
                                  n=None):
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]: I think this whole function would be much cleaner as (untested though so there might be bugs).
```
    def record_injectedparticles(self, species, w, E_total=None):
        """Handles transforming raw particle information to the information
        used to record particles as a function of time. Also handles parallel
        sum and appending to the data array the current amount of injection.

        Note:
            Assumes the fixed form of fields given in Injector().  Doesn't
            check since this is called many times.
            Since a parallelsum is performed, call this with only the species
            argument if no particles are being added by this processor.

        Arguments:
            species: [TODO - adapt to WarpX]
            w (np.ndarray): Array of length npart with particle weights.
            E_total (np.ndarray): Array of length npart with E_total values.
            n (int): Number of macroparticles, _only_ needed if overriding the
                length of arrays This is useful mostly in the case that
                w and E_total are already summed over particles, in which case a
                single number can be passed for it rather than an array.
        """
        data = np.zeros(7)

        data[0] = mwxrun.get_it() * mwxrun.get_dt()
        data[1] = mwxrun.get_it()
        data[2] = species.species_number
        data[3] = self.getvoltage_e()
        data[4] = n if np.zise(w) == 1 else np.zise(w)
        data[5] = species.sq * np.sum(w)

        if E_total is not None:
              data[6] = np.sum(E_total)

        self.append_injectedparticles(data)
```

File: mewarpx/mewarpx/emission.py [hunk 8]
```diff
@@ -127,34 +128,30 @@ def record_injectedparticles(self, species, E_total=None, w=None,
         Arguments:
             species: [TODO - adapt to WarpX]
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]: Reminder to fix.

File: mewarpx/mewarpx/emission.py [hunk 9]
```diff
@@ -177,7 +174,19 @@ def get_injectedparticles(self, clear=False):
                 originally passed field strings for lost particles. Values are
                 an (n)-shape numpy array for each field.
         """
-        raise NotImplementedError
+        lpdata = self._injectedparticles_data.data()
+
+        # Sum all except t/step/jsid/V_e from all processors
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]: Update comment (jsid is not used here).

File: mewarpx/mewarpx/emission.py [hunk 10]
```diff
@@ -414,7 +423,7 @@ def inject_particles(self):
             self.record_injectedparticles(
                 self.species,
                 particles_dict['E_total'],
-                particles_dict.get('w', None)
+                particles_dict['w']
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/emission.py]: With above change to `record_injectedparticles` the argument order would change.

File: mewarpx/mewarpx/diags_store/flux_diagnostic.py [hunk 11]
```diff
@@ -0,0 +1,915 @@
...
+    """Handles writing out charge injection and absorption."""
+
+    def __init__(self, diag_steps, scraper, runinfo,
+                 save_dir='diags/fluxes', overwrite=True,
+                 history_maxlen=5000, sig_figs=6,
+                 printed_qtys=None,
+                 check_charge_conservation=True,
+                 print_per_diagnostic=True, print_total=False,
+                 plot=True, profile_decorator=None, **kwargs):
+        """Generate and install function to write out fluxes.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/diags_store/flux_diagnostic.py:526]:
```suggestion
        """Generate and install function to write out fluxes.
        
```

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
Looks great so far. I left some initial comments, mostly minor things.

File: mewarpx/mewarpx/appendablearray.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        assert self._unitshape is not None,\
+               'Only an array with a specified unitshape can be reshaped'
+        assert len(newunitshape) == len(self._unitshape),\
+               ('New unitshape must have the same number of dimensions as original'+
+                'unitshape')
```
Comment (EtherealMC-Bit) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
               ('New unitshape must have the same number of dimensions as original '+
                'unitshape')
```

**Review by EtherealMC-Bit — CHANGES_REQUESTED** [review]
Just one comment from the other PR

File: mewarpx/mewarpx/appendablearray.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
+"""Array type which can be appended to in an efficient way.
+"""
+__all__ = ['AppendableArray','DynamicHistogram','DynamicHistogramIntersect']
```
Comment (nruof) [inline mewarpx/mewarpx/appendablearray.py]: Remove this line, DynamicHistogram and DynamicHistogramIntersect are not implemented

File: mewarpx/mewarpx/appendablearray.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+  >>> a = AppendableArray(initlen=100,typecode='d')
+  Append a single unit like this
+  >>> a.append(7.)
+  or multiple units like this
+  >>> a.append(numpy.ones(5,'d'))
+  The data can be obtained by directly indexing the instance, like this
+  >>> print a[:4]
+  [ 7., 1., 1., 1.,]
+  will give the first four number appended
+  Other methods include len, data, setautobump, cleardata, reshape
```
Comment (nruof) [inline mewarpx/mewarpx/appendablearray.py]: Update doc string format so it looks like what we currently use.
<paragraph>
Arguments:
<paragraph>

File: mewarpx/mewarpx/appendablearray.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        self._array[ss] = oldarray[ss]
+
+    def __len__(self):
+        return self._datalen
+
+    def __getitem__(self,key):
+        return self.data()[key]
+
+    def __setitem__(self,key,value):
+        self.data()[key] = value
```
Comment (nruof) [inline mewarpx/mewarpx/appendablearray.py]: Hello

**Review by nruof — CHANGES_REQUESTED** [review]
Minor formatting stuff

File: mewarpx/mewarpx/emission.py [hunk 16]
```diff
@@ -125,36 +126,34 @@ def record_injectedparticles(self, species, E_total=None, w=None,
...
-        #         data[4] = n
-        #     else:
-        #         data[4] = len(E_total)
+        data = np.zeros(7)
 
-        #     if w is not None:
-        #         data[5] = species.sq * species.sw * np.sum(w)
-        #     else:
-        #         data[5] = species.sq * species.sw * data[4]
+        data[0] = mwxrun.get_it() * mwxrun.get_dt()
```
Comment (PhilMiller) [inline mewarpx/mewarpx/emission.py:147]: Please add a comment for each of these components describing what it represents. I can readily figure out what some of these calculations mean, but not all of them.

File: mewarpx/mewarpx/mwxrun.py [hunk 17]
```diff
@@ -167,6 +167,21 @@ def _set_grid_params(self):
         else:
             raise ValueError("Unrecognized type of pywarpx.picmi Grid.")
 
+    def get_domain_area(self):
+        """Return float of simulation domain area in X & Y directions or R depending
+        on geometry."""
```
Comment (PhilMiller) [inline mewarpx/mewarpx/mwxrun.py]: This looks like it's calculating a cross-sectional area? Which is equivalent to the surface area over which current is emitted/absorbed? That seems to be what the callers use it for. Please describe it here.

File: mewarpx/mewarpx/mwxrun.py [hunk 18]
```diff
@@ -167,6 +167,21 @@ def _set_grid_params(self):
...
+                    mwxrun.zmin, mwxrun.zmax]
+
+        if mwxrun.geom_str == "XZ":
+            return (pos_lims[1] - pos_lims[0])
+        elif mwxrun.geom_str == "XYZ":
+            return ((pos_lims[1] - pos_lims[0]) * (pos_lims[3] - pos_lims[2]))
+        else:
+            # TODO: implement RZ when needed
+            warnings.warn('Returning XYZ calculation for domain area')
+            return ((pos_lims[1] - pos_lims[0]) * (pos_lims[3] - pos_lims[2]))
```
Comment (PhilMiller) [inline mewarpx/mewarpx/mwxrun.py]: If I've understood correctly, isn't this simply pi*r^2?

Regardless, I'd suggest raising an 'unimplemented' exception here, rather than warning and returning an incorrect result.

File: mewarpx/mewarpx/parallel_util.py [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+        comm.Allreduce(data, result, op=op)
+    else:
+        result = comm.allreduce(data, op=op)
+
+    return result
+
+def parallelsum(a, comm=None):
+    if mwxrun.n_procs <= 1:
+        return a
+    return mpiallreduce(a, op=mpi.SUM, comm=comm)
```
Comment (PhilMiller) [inline mewarpx/mewarpx/parallel_util.py]: Missing newline at end-of-file.

File: mewarpx/mewarpx/parallel_util.py [hunk 20]
```diff
@@ -0,0 +1,27 @@
+from mpi4py import MPI as mpi
+import numpy as np
+
+
+from mewarpx.mwxrun import mwxrun
+
+
+comm_world = mpi.COMM_WORLD
```
Comment (PhilMiller) [inline mewarpx/mewarpx/parallel_util.py:8]: This is only used in the one place. Please inline it.

File: mewarpx/mewarpx/appendablearray.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+
+
+class AppendableArray:
+    """
+    Creates an array which can be appended to in an efficient manner. The object
+    keeps an internal array which is bigger than the actual data. When new data is
+    appended, it is only copied to fill in the extra space. More space is only
+    allocated when that space fills up. This saves both the allocation time, and
+    the time to copy the existing data to the new space. This code was written by
+    David P Grote (8/19/99).
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/appendablearray.py]:
```suggestion
    David P Grote for Warp (8/19/99).
```

File: mewarpx/mewarpx/diags_store/flux_diagnostic.py [hunk 22]
```diff
@@ -0,0 +1,931 @@
+from builtins import zip
+from builtins import range
+import collections
+import copy
+import logging
+import warnings
+import os
+import glob
+
+from mewarpx.mepicmi import Species
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/diags_store/flux_diagnostic.py]: Why is this import needed?

File: mewarpx/mewarpx/parallel_util.py (hunk 20 above)
Comment (KZhu-ME) [reply mewarpx/mewarpx/parallel_util.py:8]: I believe that there will be functions implemented in the future that also use comm_world, which is why it's defined globally.

File: mewarpx/mewarpx/setups_store/diode_setup.py [hunk 23]
```diff
@@ -546,14 +546,48 @@ def init_traceparticles(self):
...
+        runvars.update(self.__dict__)
+
+        injector_dict = {'cathode': self.injector}
+        surface_dict = {
+            'cathode': self.cathode,
+            'anode': self.anode,
+            # TODO: implement accgrid, absgrid, dielectric
+            # 'accgrid': self.acc_grid,
+            # 'absgrid': self.absorption_grid,
+            # 'dielectric': [self.SiO2_1, self.SiO2_2],
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/setups_store/diode_setup.py]: We can remove these lines since the diode won't ever have other grids.

File: mewarpx/mewarpx/setups_store/diode_setup.py [hunk 24]
```diff
@@ -546,14 +546,48 @@ def init_traceparticles(self):
...
+
     def init_fluxdiag(self):
-        raise NotImplementedError(
-            "Diode FluxDiag is not yet implemented in mewarpx")
         print('### Init Diode FluxDiag ###')
+        self.fluxdiag = flux_diagnostic.FluxDiagnostic(
+            diag_steps=self.DIAG_STEPS,
+            write_dir="diags/fluxes",
+            # TODO: pass scraper when scraper is added
+            scraper=None,
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/setups_store/diode_setup.py]: These lines can be removed as well since we won't actually pass a scraper object to `FluxDiagnostic` anymore.

File: mewarpx/mewarpx/mwxrun.py (hunk 18 above)
Comment (KZhu-ME) [reply mewarpx/mewarpx/mwxrun.py]: @roelof-groenewald  can explain it better than me, but I think that the surface area calculation for RZ is more complicated since you can either emit from the face or the walls of a cylinder. But it's a good idea to put in an unimplemented exception since there isn't any RZ stuff that works right now.

File: mewarpx/mewarpx/diags_store/flux_diagnostic.py [hunk 25]
```diff
@@ -0,0 +1,931 @@
...
+                 printed_qtys=None,
+                 check_charge_conservation=True,
+                 print_per_diagnostic=True, print_total=False,
+                 plot=True, profile_decorator=None, **kwargs):
+        """Generate and install function to write out fluxes.
+
+        Arguments:
+            diag_steps (int): Number of steps between each output
+            scraper (warp.ParticleScraper): The particle scraper is instructed
+                to save E_total for lost particles.
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/diags_store/flux_diagnostic.py]: We can remove the scraper from the argument list. We won't need it (or even have one) in WarpX.

File: mewarpx/mewarpx/diags_store/flux_diagnostic.py [hunk 26]
```diff
@@ -0,0 +1,931 @@
...
+        #         for surface in self.surface_dict[key]
+        #     ]
+
+        # Initialize other variables
+        # Note that at present, last_run_step is only updated for processor 0.
+        self.last_run_step = 0
+
+        # Install into WARPX
+        # TODO: Need when scraper is implemented
+        #self.scraper.add_pidtosave('E_total')
```
Comment (roelof-groenewald) [inline mewarpx/mewarpx/diags_store/flux_diagnostic.py]:
```suggestion
```

**Review by roelof-groenewald — APPROVED** [review]
Looks good. Great work!

**Review by nruof — APPROVED** [review]
